encoding/types.mbt

```diff
@@ -0,0 +1,133 @@
+// Copyright 2024 International Digital Economy Academy
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+///|
+typealias Cont = (Decoder) -> Decode
+
+///|
+pub(all) enum Encoding {
+  UTF8
+  UTF16 // alias of UTF16LE
+  UTF16LE
+  UTF16BE
+}
+
+// Decoder
+
+///|
+priv struct Decoder {
+  mut i : @buffer.T
+  mut i_pos : Int
+  mut i_max : Int
+  t : @buffer.T
+  mut t_len : Int
+  mut t_need : Int
+  mut k : Cont
+}
+
+///|
+priv enum Decode {
+  End
+  Malformed(String)
+  Uchar(Char)
+}
+
+///|
+fn malformed(buf : @buffer.T, offset : Int, length : Int) -> Decode {
+  Malformed(buf.to_unchecked_string(offset~, length~))
+}
+
+///|
+fn malformed_pair(
+  be : Bool,
+  hi : Int,
+  buf : @buffer.T,
+  offset : Int,
+  length : Int
+) -> Decode {
+  let bs1 = buf.to_unchecked_string(offset~, length~).to_bytes()
+  let bs0 = @buffer.new(size_hint=2)
+  let (j0, j1) = if be { (0, 1) } else { (1, 0) }
+  bs0[j0] = (hi >> 8).to_byte()
+  bs0[j1] = hi.land(0xFF).to_byte()
```

Comment:
Use `bs0.write_byte()` instead of random access? The op_set method for buffer is weird